canvas.go

```diff
@@ -48,8 +52,32 @@ type line struct {
 	p0, p1 image.Point
 }
 
+var (
+	pathSanitizer = regexp.MustCompile(`\W+`)
+	piSanitizer   = regexp.MustCompile(`<\?xml[^>]*>`)
+)
+
+// sanitizeCharmPath ensures that a given string will be safe to use as a
+// CSS selector attribute (id or class).
+func sanitizeSelector(selector string) string {
+	return pathSanitizer.ReplaceAllString(selector, "-")
+}
+
+// dePI removes any processing instructions to ensure a valid SVG.
+func dePI(svg string) string {
+	return piSanitizer.ReplaceAllString(svg, "")
+}
+
 // definition creates any necessary defs that can be used later in the SVG.
-func (s *service) definition(canvas *svg.SVG) {
+func (s *service) definition(canvas *svg.SVG, iconsRendered map[string]bool) {
+	if _, ok := iconsRendered[s.charmPath]; s.iconSrc != "" && ok == false {
+		iconsRendered[s.charmPath] = true
+
+		canvas.Group(fmt.Sprintf(`id="icon-%s"`, sanitizeSelector(s.charmPath)))
```

Comment:
Rather than using sanitizeSelector, which seems potentially problematic as it will lose slash separators, conflating cs:precise/wordpress with cs:precise-wordpress, maybe
it might be nicer to just generate ids dynamically (c1, c2, c3, etc) and maintain a map from charm path to generated id?